### DISCUSSION TITLE
What constitutes good runtime parser performance?

### DISCUSSION THREAD

**Discussion by michaelfortunato** [tree-sitter/tree-sitter#3413] [Q&A] [answered]
I have written a tree-sitter grammar which generates a parser than takes ~80ms to parse my language's 6000 line file. 

Using the files attached here, https://github.com/nvim-treesitter/nvim-treesitter-context/issues/220, it took my machine 120ms to parse combined-20k.py, so python's performance is better than my parser from this one off, but they are comparable. 

Are these the parse times I should when parsing general programming languages of these file sizes? Is there any CI/CD code on the major repos that track parser runtime performance?

Thanks!

**Comment by amaanq** [comment, marked as answer]
it's not exactly easy to predict what makes some parsers faster than others (e.g. machine specs/good cpu caching can play a huge role sometimes), but there are a few numbers you can eye to ensure parser performance does not degrade a lot

1. check the state count and large state count in parser.c, the lower the better
2. ensure your external scanner is fast, if any, don't perform anything computation-heavy, it's meant to be quick

**Reply by sogaiu** [reply]
May be the second point is worth adding to [this page](https://github.com/tree-sitter/tree-sitter/wiki/Tips-and-Tricks-for-a-grammar-author)?
